agent/consul/discoverychain/compile.go

```diff
@@ -0,0 +1,489 @@
+package discoverychain
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// TODO(rb): surface any specific errors that may matter during graph vetting at write-time
+
+// Compile assembles a discovery chain from raw config entries and local context.
+func Compile(
+	serviceName string,
+	currentNamespace string,
+	currentDatacenter string,
+	entries *structs.DiscoveryChainConfigEntries,
+) (*structs.CompiledDiscoveryChain, error) {
+	if serviceName == "" {
+		return nil, fmt.Errorf("serviceName is required")
+	}
+	if currentNamespace == "" {
+		return nil, fmt.Errorf("currentNamespace is required")
+	}
+	if currentDatacenter == "" {
+		return nil, fmt.Errorf("currentDatacenter is required")
+	}
+	if entries == nil {
+		return nil, fmt.Errorf("entries is required")
+	}
+
+	c := &compiler{
+		serviceName:       serviceName,
+		currentNamespace:  currentNamespace,
+		currentDatacenter: currentDatacenter,
+		entries:           entries,
+
+		splitterNodes:      make(map[string]*structs.DiscoveryNode),
+		groupResolverNodes: make(map[structs.DiscoveryTarget]*structs.DiscoveryNode),
+		catalogQueryNodes:  make(map[structs.DiscoveryTarget]*structs.DiscoveryNode),
+
+		resolvers:       make(map[string]*structs.ServiceResolverConfigEntry),
+		retainResolvers: make(map[string]struct{}),
+		targets:         make(map[structs.DiscoveryTarget]struct{}),
+	}
+
+	// Clone this resolver map to avoid mutating the input map during compilation.
+	if len(entries.Resolvers) > 0 {
+		for k, v := range entries.Resolvers {
+			c.resolvers[k] = v
+		}
+	}
+
+	return c.compile()
+}
+
+// compiler is a single-use struct for handling intermediate state necessary
+// for assembling a discovery chain from raw config entries.
+type compiler struct {
+	serviceName       string
+	currentNamespace  string
+	currentDatacenter string
+
+	// config entries that are being compiled (will be mutated during compilation)
+	//
+	// This is an INPUT field.
+	entries *structs.DiscoveryChainConfigEntries
+
+	// cached nodes
+	splitterNodes      map[string]*structs.DiscoveryNode
+	groupResolverNodes map[structs.DiscoveryTarget]*structs.DiscoveryNode // this is also an OUTPUT field
+	catalogQueryNodes  map[structs.DiscoveryTarget]*structs.DiscoveryNode
+
+	// topNode is computed inside of assembleChain()
+	//
+	// This is an OUTPUT field.
+	topNode *structs.DiscoveryNode
```

Comment:
After reading more renaming everything would be a nightmare so I don't recommend that. maybe just updating the comments in the file to set the context that this is resolving a graph and that these variables and types correspond to "graph nodes" rather than Consul Nodes.